src/ray/gcs/gcs_client/service_based_accessor.cc

```diff
@@ -706,6 +706,22 @@ void ServiceBasedNodeInfoAccessor::AsyncReReportHeartbeat() {
   }
 }
 
+Status ServiceBasedNodeInfoAccessor::AsyncGetAllHeartbeat(
+    const MultiItemCallback<rpc::HeartbeatTableData> &callback) {
```

Comment:
I think what you want to return is `HeartbeatBatchTableData`, which includes the heartbeats and also resource_load_by_shape and placement_group_load. That would fix @rkooo567 's other comment

Reply:
That proto is generated by SendBatchedHeartbeat(), maybe you can modify that to save the latest published batch for fetching.